spec/datadog/core/configuration/components_spec.rb

```diff
@@ -263,6 +270,41 @@
           end
         end
       end
+
+      context 'with :agentless_enabled true' do
+        let(:agentless_enabled) { true }
+        let(:transport) { instance_double(Datadog::Core::Telemetry::Http::Transport) }
+        let(:expected_options) do
+          { enabled: enabled, http_transport: transport,
+            metrics_enabled: metrics_enabled, heartbeat_interval_seconds: heartbeat_interval_seconds,
+            metrics_aggregation_interval_seconds: metrics_aggregation_interval_seconds,
+            dependency_collection: dependency_collection, shutdown_timeout_seconds: shutdown_timeout_seconds }
+        end
+
+        before do
+          expect(Datadog::Core::Telemetry::Http::Transport).to receive(:build_agentless_transport).with(
+            api_key: api_key,
+            dd_site: settings.site
+          ).and_return(transport)
+        end
+
+        it { is_expected.to be(telemetry) }
+
+        context 'and no api key' do
+          let(:api_key) { nil }
+          let(:expected_options) do
+            { enabled: false, http_transport: transport,
+              metrics_enabled: false, heartbeat_interval_seconds: heartbeat_interval_seconds,
+              metrics_aggregation_interval_seconds: metrics_aggregation_interval_seconds,
+              dependency_collection: dependency_collection, shutdown_timeout_seconds: shutdown_timeout_seconds }
+          end
+
+          it 'does not enable telemetry when agentless mode requested but api key is not present' do
+            expect(logger).to receive(:debug)
+            is_expected.to be(telemetry)
+          end
+        end
```

Comment:
Minor: So this made me think the current logic is kinda confusing. If telemetry gets disabled, why are we still creating transports and the telemetry component at all? :eyes:

Reply:
To avoid null checks all over the code I would say! we create transport so that we could tell typechecker that transport is always there

Reply:
I see. I guess I'm not a big fan of the way we've implemented the pattern (and we do this in a few other places) -- I think if we need a "null object", it's probably better to introduce an actual empty implementation, rather than building the whole thing, and then having `enabled` just make it not "go" when it gets called.

(Out of curiosity, the profiler component builder does return `nil` when profiling is disabled)

Reply:
Fair enough :sweat_smile: xD